src/lxcfs.c

```diff
@@ -199,13 +232,19 @@ static void do_reload(bool reinit)
 
         dlopen_handle = dlopen(lxcfs_lib_path, RTLD_LAZY);
 	if (!dlopen_handle)
-		log_exit("%s - Failed to open liblxcfs.so", dlerror());
+		log_exit("%s - Failed to open liblxcfs.so at %s", dlerror(), lxcfs_lib_path);
 	else
 		lxcfs_debug("Opened %s", lxcfs_lib_path);
 
 good:
+        if (strlen(runtime_path) > 0 && !do_set_runtime_path(runtime_path)) {
```

Comment:
Do we really have to introduce two new exported symbols `set_runtime_path` and `lxcfslib_init` in there?

If you changing existing constructor to be a normal exported function you can also add an argument to it like this:
```
void lxcfslib_init(const char *new_path)
{
//... handle new_path here
```

As far as I understand you can't call `set_runtime_path` anywhere except *before* the `lxcfslib_init` anyways. Which is a good sign that this function should be a part of `lxcfslib_init`.

==

Second thing is that we are trying to maintain compatibility, which means that *old* LXCFS daemon should be able to load and use a new LXCFS library gracefully. With this change old LXCFS daemon will try to reload LXCFS library, constructor is missing, and it's fine and no error will be generated. But LXCFS will remain uninitialized and then it will crash or at least give fuse request failures to the user.
My idea is to move `lxcfs_init` (constructor) call to the `lxcfs_fuse_init(struct fuse_conn_info *conn, void *data)`, because it's always called almost at the same time as constructor should be called. In this case you don't need to add a new exported symbol `lxcfslib_init`. Just remove a `__attribute__((constructor))` from the `lxcfs_init` and then add an explicit call to the `lxcfs_fuse_init` for `lxcfs_init`.
Please, correct me if I'm wrong ;-)

Reply:
This solution is also not ideal because `lxcfs_fuse_init` is also relatively new (3 years, see https://github.com/lxc/lxcfs/commit/285aea407acd58811b3ecf53033497d0a25b83f9) exported symbol, which means that LXCFS daemons older than 3 years will still fail to properly load new LXCFS library. But it's better than nothing.

cc @brauner

Reply:
I looked at this for a bit, but dont see a way around breaking backwards compatibility. Essentially the problem is that we need to send a new parameter (runtime_path) between the binary and the library.

I did however merge set_runtime_path into lxcfslib_init, so there is only one new symbol.

Reply:
I guess your current solution is correct and does not break compatibility against the recent LXCFS version.

Reply:
I guess your current solution is correct and does not break compatibility against the recent LXCFS version.